### ISSUE TITLE
Appveyor / Windows environment: 'cstdint': No such file or directory

### ISSUE THREAD

**Issue by zfrenchee** [joerick/cibuildwheel#55]
Hi, 

I'm using this CIBW with travis, and it's working great! (for [this repo](https://github.com/fraenkel-lab/pcst_fast)) I tried to set up appveyor today, but ran into an issue: 
```
'cstdint': No such file or directory
```
This might be unique to 32-bit windows? If so, I'll just need to skip that build. Otherwise, if you have another suggestion, let me know!

**Comment by joerick** [comment]
`cstdint` is a C standard library header (but looks like it's being referred to by C++?). I believe cstdint is a 'newer' C header, so it may not be supported by the compiler that's on Appveyor.

You may [have some luck](https://github.com/manahl/arctic/issues/6) trying to provide the dependency yourself, or [perhaps](https://stackoverflow.com/questions/10408458/why-do-stdint-h-can-be-found-but-cstdint-not) support could be enabled by passing a compiler flag. Or maybe [this project](https://code.google.com/archive/p/msinttypes/) can help....

Might be worth a question to the Appveyor folks, but I suspect this may require a change to your code :/ note that you _do_ have to [stick to using](https://cowboyprogrammer.org/2014/06/building-python-wheels-for-windows/) Visual C++ compiler for python wheels, since that's the compiler used to compile Python itself.

**Comment by YannickJadoul** [comment]
Actually, I believe `cstdint` is the C++ version of the `stdint.h` C header, no?

But you have the pybind11 includes twice (`get_pybind_include()` and `get_pybind_include(user=True)`)? So that already seems rather weird and might be a problem?

**Comment by zfrenchee** [comment]
@YannickJadoul @joerick thanks for getting back to me. 

cstdint is a C++11 library, but I was surprised appveyor's compiler doesn't support it. Is it possible this might be an issue unique to the 32-bit windows? If I could build just for 64-bit windows would suffice. I can't tell, but it [doesn't seem like appveyor *attempted* to build the package for 64-bit windows](https://ci.appveyor.com/project/zfrenchee/pcst-fast/build/1.0.1). 

To recap, the best suggestion we have so far is that I try to provide the dependency myself?

**Comment by YannickJadoul** [comment]
This is actually rather weird, as I manage to just build a pybind11 project with MSVC on AppVeyor. The main difference is that I use VS2017.

@joerick Is this maybe just a result of the [run_with_env.cmd](https://raw.githubusercontent.com/ogrisel/python-appveyor-demo/09a1c8672e5015a74d8f69d07add6ee803c176ec/appveyor/run_with_env.cmd) script, since this swaps in a different and old (probably pre-C++11) compiler for some compilations?

**Comment by zfrenchee** [comment]
@joerick -- any ideas for what I should do here?

**Comment by joerick** [comment]
Sorry I can’t be more help, I think this is an issue intrinsic to building cross platform wheels, not an issue with cibuildwheel itself. 

The compiler that is used for older python versions is required for them to work with the Python.org distributions of Python. 

I’ve no more ideas other than the options presented above. 

Let us know what you’ve tried and we may be able to help more after you’ve done some research.

**Comment by YannickJadoul** [comment]
@zfrenchee Have you tried running this with VS2017?
Indeed, as @joerick says, it is said you need the old version of the MSVC compiler that was used to compile that Python version, but ... my project seems to be running fine when I compile everything with VS2017 (tested on a handful of different Windows machines; for a few of them I needed to install the redistributable C/C++ libraries from Windows).

**Comment by zfrenchee** [comment]
@YannickJadoul I updated the Appveyor config to use VS2017, and [the build failed again with the same error message](https://ci.appveyor.com/project/zfrenchee/pcst-fast/build/1.0.2)

**Comment by YannickJadoul** [comment]
@zfrenchee Where is this? Because in the [appveyor.yml file from that commit you have linked to](https://github.com/fraenkel-lab/pcst_fast/blob/f520a10dbedc322be7c69c15584667c8777f0fb4/appveyor.yml), I still don't see the `image: Visual Studio 2017` line in that `appveyor.yml`?

**Comment by zfrenchee** [comment]
@YannickJadoul I changed a setting on the appveyor site (SETTINGS > Environment > Build worker image)

I just added the `image: Visual Studio 2017` to my appveyor.yml file as well though, and tried a build. It seemed to fail, [but for different reasons](https://ci.appveyor.com/project/zfrenchee/pcst-fast/build/1.0.3), I think the error now is: 
```
  error: Microsoft Visual C++ 9.0 is required. 
```
You would think that would already be installed, no?

Some quick research makes it seem like fixing this might be ... _involved_:
- https://github.com/scikit-build/scikit-build/issues/216
- https://github.com/scikit-build/scikit-build/issues/34

What do you think? Should I give up?

**Comment by YannickJadoul** [comment]
@zfrenchee Apologies, I'd forgotten about the settings UI of AppVeyor.

> You would think that would already be installed, no?

On https://www.appveyor.com/docs/build-environment/#python, it says that `Visual C++ Compiler for Python 2.7` is not part of the VS2017 image, somehow.

Weirdly enough, my build setup doesn't seem to care about that and just builds the wheel anyway. So I'm guessing this check for MSVC 9.0 is part of the `distutils` library (and since I'm just firing up CMake through `setup.py`, that check is not happening for my builds)? But in the end, this is just the old 'building wheels for Windows' problem again, it seems. Because I'm guessing that even when you're not using `cibuildwheel` you'll run into errors?

**Comment by zfrenchee** [comment]
@YannickJadoul Has anything changed here? I'm just re-building my package today and uploading wheels to pypi, but still don't have a strategy for getting Windows wheels, which makes some of my colleagues sad (cc @tobieh).

**Comment by joerick** [comment]
@zfrenchee If you can get your wheels to build on appveyor with `python setup.py bdist_wheel`, (without cibuildwheel) that would give us an idea where the problem might lie! At the moment I can't tell if this is a problem between your code and the compiler or if cibuildwheel is involved.

**Comment by zfrenchee** [comment]
@YannickJadoul could you link the appveyor.yml you're referring to? I'd love to model [mine](https://github.com/fraenkel-lab/pcst_fast/blob/master/appveyor.yml) off yours. 

My project is https://github.com/fraenkel-lab/pcst_fast and I have a [makefile](https://github.com/fraenkel-lab/pcst_fast/blob/master/Makefile) as well. I'm agnostic as to how the windows build is built, I'd just love to see it up and running. I have at least 2 windows users who are :( about the lack of a windows wheel.

**Comment by YannickJadoul** [comment]
@zfrenchee Sure, the appveyor.yml is [here](https://github.com/YannickJadoul/Parselmouth/blob/master/.appveyor.yml), but I'm guess you also/mostly will want to take a look at the [setup.py](https://github.com/YannickJadoul/Parselmouth/blob/master/setup.py).

By the way, could you also try without using cibuildwheel, and just try running a few cases such as `C:\Python27\python.exe setup.py build` and see what happens? If that actually works and doesn't give you an error, we might see what happens if we do not always run this `run_with_env.cmd`...

**Comment by YannickJadoul** [comment]
By the way, you might be interested in this section of the pybind11 documentation, reassuring you that this approach should not be too problematic: https://pybind11.readthedocs.io/en/stable/faq.html#working-with-ancient-visual-studio-2009-builds-on-windows
